### PR TITLE
add handle for custom device

### PR DESCRIPTION
**Pull request by smazouz42**
This pull request addresses issue https://github.com/pyccel/pyccel-cuda/issues/41 by implementing a new feature in Pyccel that allows users to define a custom device

**Commit Summary**

- Adding handler for custom device and its code generation.
- Adding test

### CODE REVIEW

**Review by jalalium — APPROVED** [review]
Very good job!

File: docs/cuda.md [hunk 1]
```diff
@@ -20,4 +20,27 @@ threadsperblock = 1
...
-```
\ No newline at end of file
+```
+
+### device
+
+Device functions are similar to kernels, but are executed within the context of a kernel. They can be called only from kernels or device functions, and are typically used for operations that are too small to justify launching a separate kernel, or for operations that need to be performed repeatedly within the context of a kernel.
+
+```python
+from pyccel.decorators import device , kernel
```
Comment (EmilyBourne) [inline docs/cuda.md:30]:
```suggestion
from pyccel.decorators import device, kernel
```

File: pyccel/parser/semantic.py [hunk 2]
```diff
@@ -136,9 +136,10 @@
         UNUSED_DECORATORS, UNSUPPORTED_POINTER_RETURN_VALUE, PYCCEL_RESTRICTION_OPTIONAL_NONE,
         PYCCEL_RESTRICTION_PRIMITIVE_IMMUTABLE, PYCCEL_RESTRICTION_IS_ISNOT,
         FOUND_DUPLICATED_IMPORT, UNDEFINED_WITH_ACCESS, MACRO_MISSING_HEADER_OR_FUNC, PYCCEL_RESTRICTION_INHOMOG_SET,
-        MISSING_KERNEL_CONFIGURATION,
+        MISSING_KERNEL_CONFIGURATION,INVAlID_DEVICE_CALL,
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:139]:
```suggestion
        MISSING_KERNEL_CONFIGURATION, INVAlID_DEVICE_CALL,
```

**Review by EmilyBourne — COMMENTED** [review]
This seems like a good implementation of the described issue. Good work. I only have minor cosmetic comments

File: pyccel/parser/semantic.py (hunk 2 above)
Comment (smazouz42) [reply pyccel/parser/semantic.py]: [bc11b14](https://github.com/pyccel/pyccel-cuda/pull/61/commits/bc11b147863d91f8cbc66d2a3c47d34872d258aa)

File: docs/cuda.md (hunk 1 above)
Comment (smazouz42) [reply docs/cuda.md]: [bc11b14](https://github.com/pyccel/pyccel-cuda/pull/61/commits/bc11b147863d91f8cbc66d2a3c47d34872d258aa)

**Review by EmilyBourne — APPROVED** [review]

**Pull request merged**